enhancements/update/versioning-cincinnati-api.md

```diff
@@ -0,0 +1,122 @@
+Embed the version information in json schema. The updated json schema should look like:
+```json
+{
+    "version": 1,
+    "nodes": [],
+    "edges": []
+}
+```
+
+#### Media Type Versioning
+Currently, the versioning is done by using the versioned URI and does not support media type versioning. The user should
+be able to request a specific version of cincinnati graph by supplying the version in media type field of the request.
+for example `application/vnd.redhat.cincinnati.v1+json`. Here are some of the
+[examples of media type versioning](https://github.com/opencontainers/image-spec/blob/v1.0.1/media-types.md).  
+In the event where the media type is not provided, or `application/json` is provided, cincinnati should return a default
+response that is compatible with the oldest supported version so as not to break old clusters from upgrading.
+
+
+#### URI Change
+The current URI follows `{URL}/v1/graph`. The new un-versioned URI will follow `{URL}/graph`. The existing URI can be
+redirected to the new URI. This can be done through cincinnati. In the case where an old URI is being called, cincinnati
+will by default return the graph that is compatible with the oldest supported version.
+
+
+#### Version metrics
+Cincinnati should also implement a way to understand how many clients are consuming the various versions.
+This can be done by implementing metrics for old URI vs new URI in Cincinnati. Similarly, metrics should be implemented
+to understand the consumption of various media types (versions). This will help us to make decisions on deprecating the
+old versions and URI. For example, a single `cincinnati_pe_http_response_total` counter with labels for `uri` and `version`
+
+
+### Risks and Mitigations
+* As there are some customers that have old URI hardcoded while creating the cluster, we will have to support old URI
+  for indefinite period of time. We can redirect the old URI to new one to avoid maintaining different URIs, but we'll
+  still have to maintain different versions and v1 for indefinite period of time.
+
+
+## Design Details
+
+### Test Plan
+* cincinnati tests to make sure we are not breaking older versions.
+* cincinnati returns correct resource version when the content type is provided.
+
+### Graduation Criteria
+GA. When it works, we ship it.
+
+
+### Version Skew Strategy
+* The `v1` URI will be supported for foreseeable future, this applies only to URI.
+* Maintenance cut-offs can be set for versions. Also, developers can take decision on deprecating older versions
+  depending upon the version metrics gathered. Although `v1` URI will be supported for foreseeable future, the
+  individual versions will not.
+
+## Implementation History
+
+## Drawbacks
+
+
+## Alternatives
+* We can continue with the current URI based versioning for the API and adding version information in the json schema.  
+  **Drawback**: For every minor non-breaking change with a bumped version, the client will have to change the URI with
```

Comment:
Since the default is to return the oldest supported format, won't the client have to explicitly request the new version anyway?

Reply:
Yes.  Can we add subsections here to go into more detail?  Something like:

```markdown
## Alternatives

### URI versioning

In addition to adding version information to the JSON Schema for downstream consumers, we can continue with the current URI based versioning for the Cincinnati API.
The benefit to that approach is that we don't need to deprecate the existing v1 URI.
A downside is that clients need to either:

* Poll a series of versioned URIs until they find one the service supports.
  For example, the client could poll the v3 URI, get a 404, poll the v2 URI, get a 404, and then successfully hit the v1 URI if the client preferred v3, but the service they were hitting only supported v1.
* Hit a new URI that lists versions supported by the service.
  For example, `/graph/versions`, which would return "I serve v1 at `$URI_1`, v2 at `$URI_2`, etc.", from which the client selects its favorite.
  The HTTP RFCs call this [reactive negotiation][1].

While these alternatives would work, it's simpler in the client code to construct an `Accept` header and make a single request, and one of the cluster-version operator's design goals is to make the client as simple as possible, because bugs in the cluster-version operator can be hard to recover from.

### Media-type versioning with the existing, versioned URI

We could serve multiple, media-typed versions from the existing add content-type negotiation with the existing `{URL}/v1/graph` location.
The benefit to that approach is that we don't need to deprecate the existing URI.
The drawback is that users might find the dissonance confusing while reading docs talking about v2 content from a v1 URI.

[1]: https://datatracker.ietf.org/doc/html/rfc7231#section-3.4.2
```

Reply:
This sounds good to me.